tiatoolbox/models/engine/engine_abc.py

```diff
@@ -1,75 +1,165 @@
 from abc import ABC, abstractmethod
-from typing import List, Tuple, Union
 
-import numpy as np
+import torch.nn as nn
 
+from tiatoolbox.models.architecture import get_pretrained_model
 
-class IOConfigABC(ABC):
-    """Define an abstract class for holding predictor I/O information.
 
-    Enforcing such that following attributes must always be defined by
-    the subclass.
+class EngineABC(ABC):
+    """Abstract base class for engines used in tiatoolbox.
+
+    Args:
+        model (nn.Module):
+            Use externally defined PyTorch model for prediction with.
+            weights already loaded. Default is `None`. If provided,
+            `pretrained_model` argument is ignored.
+        pretrained_model (str):
+            Name of the existing models support by tiatoolbox for
+            processing the data. For a full list of pretrained models,
+            refer to the `docs
+            <https://tia-toolbox.readthedocs.io/en/latest/pretrained.html>`_
+            By default, the corresponding pretrained weights will also
+            be downloaded. However, you can override with your own set
+            of weights via the `pretrained_weights` argument. Argument
+            is case-insensitive.
+        pretrained_weights (str):
+            Path to the weight of the corresponding `pretrained_model`.
+
+            >>> engine = EngineABC(
+            ...    pretrained_model="pretrained-model-name",
+            ...    pretrained_weights="pretrained-local-weights.pth")
+
+        batch_size (int):
+            Number of images fed into the model each time.
+        num_loader_workers (int):
+            Number of workers to load the data using :class:`torch.utils.data.Dataset`.
+            Please note that they will also perform preprocessing. default = 0
+        num_postproc_workers (int):
+            Number of workers to postprocess the results of the model. default = 0
+        verbose (bool):
+            Whether to output logging information.
+
+    Attributes:
+        images (str or :obj:`pathlib.Path` or :obj:`numpy.ndarray`):
+            A HWC image or a path to WSI.
+        mode (str):
```

Comment:
```suggestion
        mode (str | None):
```

if WSIReader can identify automatically otherwise it can be manually specified for confusing formats such as "tiff".